File: libs/document-processor-next/src/engine/checkpoint-manager.ts
import { Blue, BlueNode, type Node } from '@blue-labs/language';

import { KEY_CHECKPOINT } from '../constants/processor-contract-constants.js';
import {
  relativeCheckpointLastEvent,
  relativeCheckpointLastSignature,
  RELATIVE_CHECKPOINT,
} from '../constants/processor-pointer-constants.js';
import { resolvePointer } from '../util/pointer-utils.js';
import type { ContractBundle } from './contract-bundle.js';
import type { ChannelEventCheckpoint, MarkerContract } from '../model/index.js';
import { DocumentProcessingRuntime } from '../runtime/document-processing-runtime.js';

const blueHelper = new Blue();

function createEmptyCheckpointNode(): Node {
  return blueHelper.jsonValueToNode({
    type: { blueId: 'ChannelEventCheckpoint' },
    lastEvents: {},
    lastSignatures: {},
  });
}

function createEmptyCheckpointContract(): ChannelEventCheckpoint {
  return {
    key: KEY_CHECKPOINT,
    lastEvents: {},
    lastSignatures: {},
  };
}

function isChannelEventCheckpoint(
  marker: MarkerContract,
): marker is ChannelEventCheckpoint {
  return (
    marker != null &&
    Object.prototype.hasOwnProperty.call(marker, 'lastEvents') &&
    Object.prototype.hasOwnProperty.call(marker, 'lastSignatures')
  );
}

export class CheckpointRecord {
  lastEventNode: Node | null;
  lastEventSignature: string | null;

  constructor(
    readonly markerKey: string,
    readonly checkpoint: ChannelEventCheckpoint,
    readonly channelKey: string,
    lastEventNode: Node | null,
    lastEventSignature: string | null,
  ) {
    this.lastEventNode = lastEventNode;
    this.lastEventSignature = lastEventSignature;
  }

  matches(signature: string | null | undefined): boolean {
    return signature != null && signature === this.lastEventSignature;
  }
}

export class CheckpointManager {
  constructor(
    private readonly runtime: DocumentProcessingRuntime,
    private readonly signatureFn: (node: Node | null) => string | null,
  ) {}

  ensureCheckpointMarker(scopePath: string, bundle: ContractBundle): void {
    const marker = bundle.marker(KEY_CHECKPOINT);
    const pointer = resolvePointer(scopePath, RELATIVE_CHECKPOINT);
    if (!marker) {
      const markerNode = createEmptyCheckpointNode();
      this.runtime.directWrite(pointer, markerNode);
      bundle.registerCheckpointMarker(createEmptyCheckpointContract());
      return;
    }
    if (!isChannelEventCheckpoint(marker)) {
      throw new Error(
        `Reserved key 'checkpoint' must contain a Channel Event Checkpoint at ${pointer}`,
      );
    }
  }

  findCheckpoint(bundle: ContractBundle, channelKey: string): CheckpointRecord | null {
    for (const [markerKey, marker] of bundle.markerEntries()) {
      if (!isChannelEventCheckpoint(marker)) {
        continue;
      }
      const stored = marker.lastEvents?.[channelKey] ?? null;
      const storedClone = stored ? stored.clone() : null;
      const storedSignature = marker.lastSignatures?.[channelKey] ?? null;
      const signature = storedSignature ?? this.signatureFn(storedClone);
      return new CheckpointRecord(markerKey, marker, channelKey, storedClone, signature);
    }
    return null;
  }

  isDuplicate(record: CheckpointRecord | null, signature: string | null | undefined): boolean {
    return record != null && record.matches(signature);
  }

  persist(
    scopePath: string,
    bundle: ContractBundle,
    record: CheckpointRecord | null,
    eventSignature: string | null,
    eventNode: Node | null,
  ): void {
    if (!record) {
      return;
    }
    const eventPointer = resolvePointer(
      scopePath,
      relativeCheckpointLastEvent(record.markerKey, record.channelKey),
    );
    const stored = eventNode ? eventNode.clone() : null;
    this.runtime.chargeCheckpointUpdate();
    this.runtime.directWrite(eventPointer, stored);

    if (!record.checkpoint.lastEvents) {
      record.checkpoint.lastEvents = {};
    }
    if (stored) {
      record.checkpoint.lastEvents[record.channelKey] = stored.clone();
    } else {
      delete record.checkpoint.lastEvents[record.channelKey];
    }
    record.lastEventNode = stored ? stored.clone() : null;

    const signaturePointer = resolvePointer(
      scopePath,
      relativeCheckpointLastSignature(record.markerKey, record.channelKey),
    );
    const signatureNode = eventSignature == null
      ? null
      : new BlueNode().setValue(eventSignature);
    this.runtime.directWrite(signaturePointer, signatureNode);

    if (!record.checkpoint.lastSignatures) {
      record.checkpoint.lastSignatures = {};
    }
    if (eventSignature == null) {
      delete record.checkpoint.lastSignatures[record.channelKey];
    } else {
      record.checkpoint.lastSignatures[record.channelKey] = eventSignature;
    }
    record.lastEventSignature = eventSignature ?? null;
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/CheckpointManager.java
package blue.language.processor;

import blue.language.model.Node;
import blue.language.processor.model.ChannelEventCheckpoint;
import blue.language.processor.model.MarkerContract;
import blue.language.processor.util.PointerUtils;
import blue.language.processor.util.ProcessorContractConstants;
import blue.language.processor.util.ProcessorPointerConstants;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;

/**
 * Handles per-scope checkpoint lifecycle: lazy creation, gating, and persistence.
 */
final class CheckpointManager {

    private final DocumentProcessingRuntime runtime;
    private final Function<Node, String> signatureFn;

    CheckpointManager(DocumentProcessingRuntime runtime, Function<Node, String> signatureFn) {
        this.runtime = Objects.requireNonNull(runtime, "runtime");
        this.signatureFn = Objects.requireNonNull(signatureFn, "signatureFn");
    }

    void ensureCheckpointMarker(String scopePath, ContractBundle bundle) {
        MarkerContract marker = bundle.marker(ProcessorContractConstants.KEY_CHECKPOINT);
        String pointer = PointerUtils.resolvePointer(scopePath, ProcessorPointerConstants.RELATIVE_CHECKPOINT);
        if (marker == null) {
            Node markerNode = new Node()
                    .type(new Node().blueId("ChannelEventCheckpoint"))
                    .properties("lastEvents", new Node().properties(new LinkedHashMap<>()))
                    .properties("lastSignatures", new Node().properties(new LinkedHashMap<>()));
            runtime.directWrite(pointer, markerNode);
            bundle.registerCheckpointMarker(new ChannelEventCheckpoint());
            return;
        }
        if (!(marker instanceof ChannelEventCheckpoint)) {
            throw new IllegalStateException(
                    "Reserved key 'checkpoint' must contain a Channel Event Checkpoint at " + pointer);
        }
    }

    CheckpointRecord findCheckpoint(ContractBundle bundle, String channelKey) {
        for (Map.Entry<String, MarkerContract> entry : bundle.markerEntries()) {
            if (entry.getValue() instanceof ChannelEventCheckpoint) {
                ChannelEventCheckpoint checkpoint = (ChannelEventCheckpoint) entry.getValue();
                Node stored = checkpoint.lastEvent(channelKey);
                CheckpointRecord record = new CheckpointRecord(entry.getKey(), checkpoint, channelKey, stored);
                String storedSignature = checkpoint.lastSignature(channelKey);
                record.lastEventSignature = storedSignature != null ? storedSignature : signatureFn.apply(stored);
                return record;
            }
        }
        return null;
    }

    boolean isDuplicate(CheckpointRecord record, String signature) {
        return record != null && record.matches(signature);
    }

    void persist(String scopePath,
                 ContractBundle bundle,
                 CheckpointRecord record,
                 String eventSignature,
                 Node eventNode) {
        if (record == null) {
            return;
        }
        String pointer = PointerUtils.resolvePointer(scopePath,
                ProcessorPointerConstants.relativeCheckpointLastEvent(record.markerKey, record.channelKey));
        Node stored = eventNode != null ? eventNode.clone() : null;
        runtime.chargeCheckpointUpdate();
        runtime.directWrite(pointer, stored);
        record.checkpoint.updateEvent(record.channelKey, stored);
        record.lastEventNode = stored != null ? stored.clone() : null;
        String signaturePointer = PointerUtils.resolvePointer(scopePath,
                ProcessorPointerConstants.relativeCheckpointLastSignature(record.markerKey, record.channelKey));
        Node signatureNode = eventSignature != null ? new Node().value(eventSignature) : null;
        runtime.directWrite(signaturePointer, signatureNode);
        record.checkpoint.updateSignature(record.channelKey, eventSignature);
        record.lastEventSignature = eventSignature;
    }

    static final class CheckpointRecord {
        final String markerKey;
        final ChannelEventCheckpoint checkpoint;
        final String channelKey;
        Node lastEventNode;
        String lastEventSignature;

        CheckpointRecord(String markerKey,
                         ChannelEventCheckpoint checkpoint,
                         String channelKey,
                         Node lastEventNode) {
            this.markerKey = markerKey;
            this.checkpoint = checkpoint;
            this.channelKey = channelKey;
            this.lastEventNode = lastEventNode != null ? lastEventNode.clone() : null;
        }

        boolean matches(String signature) {
            return signature != null && signature.equals(lastEventSignature);
        }
    }
}
